### PR TITLE
add macro-averaged precision,recall,f1 options in evaluation

### PR DESCRIPTION
**Pull request by LXXXXR**

### CODE REVIEW

File: tests/test_dataset.py [hunk 1]
```diff
@@ -61,6 +61,18 @@ def test_datasets_override_default(dataset_name):
     assert dataset.CLASSES == original_classes
 
 
+@patch.multiple(BaseDataset, __abstractmethods__=set())
+def test_dataset_evaluation():
+    dataset = BaseDataset(data_prefix='', pipeline=[], test_mode=True)
+    dataset.data_infos = [dict(gt_label=0), dict(gt_label=1)]
+    fake_results = np.array([[1, 0], [0, 1]])
```
Comment (ycxioooong) [inline tests/test_dataset.py:68]: how about more complex test samples?
For example, use 3 classes, and with repeated results?

File: mmcls/models/losses/__init__.py [hunk 2]
```diff
@@ -1,9 +1,11 @@
 from .accuracy import Accuracy, accuracy
 from .cross_entropy_loss import CrossEntropyLoss, cross_entropy
+from .f1 import f_1, precision, recall
```
Comment (ycxioooong) [inline mmcls/models/losses/__init__.py:3]: would be better to use ``f1_score`` as the function name

File: mmcls/models/losses/f1.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    if isinstance(pred, torch.Tensor) and isinstance(target, torch.Tensor):
+        confusion_matrix = calculate_confusion_matrix_torch(pred, target)
+    elif isinstance(pred, np.ndarray) and isinstance(target, np.ndarray):
+        confusion_matrix = calculate_confusion_matrix_numpy(pred, target)
+    else:
+        raise TypeError('pred and target should both be'
+                        'torch.Tensor or np.ndarray')
+
+    with torch.no_grad():
+        res = confusion_matrix.diag() / confusion_matrix.sum(0)
```
Comment (ycxioooong) [inline mmcls/models/losses/f1.py:73]: use xxx/ max(a, 1) to avoid dividing by zero

File: mmcls/models/losses/f1.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
+import numpy as np
+import torch
+
```
Comment (ycxioooong) [inline mmcls/models/losses/f1.py:3]: It's better to use a more general filename.

File: mmcls/models/losses/eval_metrics.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        pred (torch.Tensor | np.array): The model prediction.
+        target (torch.Tensor | np.array): The target of each prediction.
+
+    Returns:
+        float: The function will return a single float as precision.
+    """
+    if isinstance(pred, torch.Tensor) and isinstance(target, torch.Tensor):
+        confusion_matrix = calculate_confusion_matrix_torch(pred, target)
+    elif isinstance(pred, np.ndarray) and isinstance(target, np.ndarray):
+        confusion_matrix = calculate_confusion_matrix_numpy(pred, target)
```
Comment (ZwwWayne) [inline mmcls/models/losses/eval_metrics.py:42]: I suggest we simply use one `calculate_confusion_matrix`, process different types inside the function, and convert one to another.

**Review by ycxioooong — APPROVED** [review]

**Pull request merged**